File: fl_backend/app/api/aktionen/schemas.py
from typing import Any, Literal

from bson import ObjectId
from pydantic import BaseModel, Field, TypeAdapter, field_validator

from app.shared.schemas.bounds import LIST_LIMIT_DEFAULT, LIST_LIMIT_MAX
from app.shared.schemas.custom import CustomObjectId
from app.shared.schemas.responses import BaseAPIResponse

# One alias for both the row and the filter: without it the two hand-copies drift, and a filter
# narrower than the rows answers a legal `?operation=` with a 422 that no test would catch.
FLAktionOperation = Literal["insert", "insert_many", "patch_one", "patch_many", "delete_many", "erase_many"]


def _stringify_oids(value: Any) -> Any:
    """Every `ObjectId` in a stored document, at any depth, rendered as its hex.

    Recursive because a fixture embeds ids inside `team1`, `ort` and `schiedsrichter`, so a top-level
    pass would leave the ones that actually break serialization.
    """

    if isinstance(value, ObjectId):
        return str(value)
    if isinstance(value, dict):
        return {key: _stringify_oids(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_stringify_oids(item) for item in value]

    return value


class FLAktor(BaseModel):
    """Who a write was attributed to. Mirrors `app/core/recording.py :: Actor`."""

    kind: Literal["admin_session", "system", "public"]
    email: str


class FLAktionRequest(BaseModel):
    method: str
    path: str


class FLAktion(BaseModel):
    """One recorded write.

    A read model, so nothing here refuses a stored value (`docs/backend/spec.md :: I36`): this
    collection holds copies of documents from every other one, and a row it cannot serve is a row
    nobody can see was recorded.
    """

    id: CustomObjectId = Field(validation_alias="_id", serialization_alias="id")
    at: str
    actor: FLAktor
    correlation_id: str
    request: FLAktionRequest | None
    collection: str
    operation: FLAktionOperation
    document_id: str | None
    db_filter: dict[str, str] | None
    # A list is `delete_many`'s: a removal follows no write a restore could replay, so one row
    # carries every image it took (`docs/backend/spec.md :: I48`).
    before: dict[str, Any] | list[dict[str, Any]] | None
    modified_count: int | None
    redacted_at: str | None

    @field_validator("document_id", mode="before")
    @classmethod
    def _as_text(cls, value: Any) -> str | None:
        """An ObjectId everywhere but `saisons`, whose `_id` is the season string.

        Text rather than a union: the wire names only the row a restore targets; a
        removal's id array serves the redaction alone (`docs/backend/spec.md :: I42`).
        """

        return None if value is None or isinstance(value, list) else str(value)

    @field_validator("before", mode="before")
    @classmethod
    def _json_safe(cls, value: Any) -> Any:
        """The stored image is a real Mongo document, so its ids are `ObjectId`s that JSON cannot carry.

        Converted on the way out rather than on the way in: the row keeps the document as it stood,
        and only the wire sees text (`docs/backend/spec.md :: I43`).
        """

        return _stringify_oids(value)


FLAktionenListAdapter = TypeAdapter(list[FLAktion])


class FLAktionenFilterParams(BaseModel):
    collection: str | None = None
    operation: FLAktionOperation | None = None
    correlation_id: str | None = None

    limit: int = Field(default=LIST_LIMIT_DEFAULT, ge=1, le=LIST_LIMIT_MAX)
    # Newest first, and no `sort_by`: every other order over an append-only log is a report rather
    # than a page, and one of them would have to be the default anyway.
    order: Literal["asc", "desc"] = Field(default="desc")


class FLAktionenListResponse(BaseAPIResponse):
    aktionen: list[FLAktion]

File: fl_frontend/src/features/aktionen/schemas.ts
import z from "zod";

import { BaseAPIResponseSchema } from "@/core/schemas";
import { CustomObjectIdStringSchema } from "@/shared/schemas";

export const FLAktorMitAdresseSchema = z.object({
  // `public` is a visitor with no session at all — the application form's own write. `email` is a
  // mailbox for `admin_session` alone, and a sentinel for the other two.
  kind: z.enum(["admin_session", "system", "public"]),
  email: z.string(),
});

/** A signed-in person, named by a pseudonym and the Funktion the write was authorised under, and never by an address. */
export const FLAktorPersonSchema = z.object({
  kind: z.enum(["person_session"]),
  pseudonym: z.string(),
  funktion: z.enum(["kontakt", "spieler", "schiedsrichter"]),
});
export type FLAktorPerson = z.infer<typeof FLAktorPersonSchema>;

/** On `kind`, so a person's row can never be read as carrying an address. */
export const FLAktorSchema = z.discriminatedUnion("kind", [FLAktorMitAdresseSchema, FLAktorPersonSchema]);
export type FLAktor = z.infer<typeof FLAktorSchema>;

export const FLAktionRequestSchema = z.object({
  method: z.string(),
  path: z.string(),
});
export type FLAktionRequest = z.infer<typeof FLAktionRequestSchema>;

/**
 * One recorded write. **Every field is exactly as wide as the backend's and no wider** — this collection holds copies of
 * documents from every other one, so a mirror refusing a single stored value takes the whole page down with it.
 */
export const FLAktionSchema = z.object({
  id: CustomObjectIdStringSchema,
  // Not `CustomDateStringSchema`: an instant carrying an offset, where every other date in this app is a calendar day.
  at: z.string(),
  actor: FLAktorSchema,
  trace_id: z.string(),
  // Null on a write made outside a request, which is what the system actor records.
  request: FLAktionRequestSchema.nullable(),
  // Open rather than an enum of the recorded names: a collection added on the backend must still list here.
  collection: z.string(),
  operation: z.enum(["insert", "insert_many", "patch_one", "patch_many", "delete_many", "erase_many"]),
  // An ObjectId everywhere but `saisons`, whose id is the season string. Null on a fan-out, which named a filter
  // instead, and on a bulk create, which named nothing at all (`docs/backend/spec.md :: I40`).
  document_id: z.string().nullable(),
  db_filter: z.record(z.string(), z.string()).nullable(),
  // In the replaced document's place: the list never carries an image, only whether the row secured
  // one — `GET /aktionen/{aktion_id}` is the read that serves it (`docs/backend/spec.md :: I43`).
  stand_gesichert: z.boolean(),
  modified_count: z.int().nullable(),
  // Set once a person's erasure, or a referee's anonymisation, has overwritten the values this row recorded.
  redacted_at: z.string().nullable(),
});
export type FLAktion = z.infer<typeof FLAktionSchema>;

export const FLAktionenListResponseSchema = BaseAPIResponseSchema.extend({
  aktionen: z.array(FLAktionSchema),
  /** False where the endpoint's cap cut the answer short — and a year of recorded writes reaches that cap without a flood. */
  vollstaendig: z.boolean(),
  /**
   * Open keys: a stored value no option names would take the page down rather than going uncounted, and an absent
   * key reads as zero. Counted over the rows a trace or document narrowing leaves, never the whole log.
   */
  anzahl_je_collection: z.record(z.string(), z.int().nonnegative()),
  anzahl_je_operation: z.record(z.string(), z.int().nonnegative()),
  anzahl_je_herkunft: z.record(z.string(), z.int().nonnegative()),
});
export type FLAktionenListResponse = z.infer<typeof FLAktionenListResponseSchema>;
